quickwit/quickwit-ingest/src/ingest_v2/fetch.rs

```diff
@@ -0,0 +1,605 @@
+    shard_status_rx: watch::Receiver<ShardStatus>,
+    batch_num_bytes: usize,
+}
+
+impl fmt::Debug for FetchTask {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        f.debug_struct("FetchTask")
+            .field("client_id", &self.client_id)
+            .field("index_uid", &self.index_uid)
+            .field("source_id", &self.source_id)
+            .field("shard_id", &self.shard_id)
+            .finish()
+    }
+}
+
+impl FetchTask {
+    pub const DEFAULT_BATCH_NUM_BYTES: usize = 1024 * 1024; // 1 MiB
+
+    pub fn spawn(
+        open_fetch_stream_request: OpenFetchStreamRequest,
+        mrecordlog: Arc<RwLock<MultiRecordLog>>,
+        shard_status_rx: watch::Receiver<ShardStatus>,
+        batch_num_bytes: usize,
+    ) -> ServiceStream<IngestV2Result<FetchResponseV2>> {
+        let (fetch_response_tx, fetch_stream) = ServiceStream::new_bounded(3);
+        let from_position_inclusive = open_fetch_stream_request
+            .from_position_exclusive
+            .map(|position| position + 1)
+            .unwrap_or(0);
+        let to_position_inclusive = open_fetch_stream_request
+            .to_position_inclusive
+            .unwrap_or(u64::MAX);
+        let mut fetch_task = Self {
+            queue_id: open_fetch_stream_request.queue_id(),
+            client_id: open_fetch_stream_request.client_id,
+            index_uid: open_fetch_stream_request.index_uid.into(),
+            source_id: open_fetch_stream_request.source_id,
+            shard_id: open_fetch_stream_request.shard_id,
+            from_position_inclusive,
+            to_position_inclusive,
+            mrecordlog,
+            fetch_response_tx,
+            shard_status_rx,
+            batch_num_bytes,
+        };
+        let future = async move { fetch_task.run().await };
+        tokio::spawn(future);
+        fetch_stream
+    }
+
+    async fn run(&mut self) {
+        debug!(
+            client_id=%self.client_id,
+            index_uid=%self.index_uid,
+            source_id=%self.source_id,
+            shard_id=%self.shard_id,
+            from_position_exclusive=%self.from_position_inclusive,
+            to_position_inclusive=%self.to_position_inclusive,
+            "Spawning fetch task."
+        );
+        'outer: while self.from_position_inclusive <= self.to_position_inclusive {
+            while self.shard_status_rx.borrow().replication_position_inclusive
+                < self.from_position_inclusive
+            {
+                if self.shard_status_rx.changed().await.is_err() {
+                    // The ingester was dropped.
+                    break 'outer;
+                }
+            }
+            let fetch_range = self.from_position_inclusive..=self.to_position_inclusive;
+            let mrecordlog_guard = self.mrecordlog.read().await;
+
+            let Ok(docs) = mrecordlog_guard.range(&self.queue_id, fetch_range) else {
+                // The queue no longer exists.
+                break;
```

Comment:
Is this the right behavior here?
Do we want to log a warning or something?

Reply:
A warning won't hurt. If that happens that means the grace period is too short or some clients are too slow to consume the stream.